src/java/org/apache/cassandra/service/StorageService.java

```diff
@@ -5287,12 +5287,16 @@ protected synchronized void drain(boolean isFinalShutdown) throws IOException, I
             CommitLog.instance.shutdownBlocking();
 
             // wait for miscellaneous tasks like sstable and commitlog segment deletion
-            ScheduledExecutors.nonPeriodicTasks.shutdown();
-            if (!ScheduledExecutors.nonPeriodicTasks.awaitTermination(1, MINUTES))
-                logger.warn("Unable to terminate non-periodic tasks within 1 minute.");
-
             ColumnFamilyStore.shutdownPostFlushExecutor();
-            setMode(Mode.DRAINED, !isFinalShutdown);
+
+            try
+            {
+                ScheduledExecutors.shutdownNowAndWait(1, MINUTES);
```

Comment:
Did you think through the consequences of also shutting down `scheduledFastTasks` here too? `approxTime()` will stop updating and tests like `isAfter()` will get stuck returning `false` even as real time passes.

As `drain()` can be called by `nodetool` independently of the shutdown hook, is there any functionality that will be impacted by `approxTime` sticking until the node is finally shutdown?

Reply:
I am sorry I am not getting your first paragraph at all. What approxTime? The second paragraph builds on top of me not understanding the first.

Reply:
@jonmeredith  I think I got it, I rewrote it a little bit. You will see it in the next commit. I think, to be super safe, it is better if we move the shutdown of that very specific fast tasks executor to Runtime's shutdown hooks so we shut down as the very last thing so we still progress time at least while c* is between drain and the actual shut down area.